### ISSUE TITLE
Please document that Tokio doesn't work with `fork()`

### ISSUE THREAD

**Issue by jaskij** [tokio-rs/tokio#4301]
**Version**
N/A - checked in current `master` and at `tokio-1.14.0`

**Platform**
N/A

**Description**

Over the last few days I've seen two people in the #tokio-users Discord having issues with `fork`ing. It seems the interactions between `fork()` and the runtime are not documented at all. This has shown to stumble users and as such I'm filing this as a documentation bug.

Out of curiosity I have cloned the repository and looked for mentions of `fork`:

```
jaskij@jaskij-pc [16:40:00] [~/projects/rust/tokio] [master]
-> % rg -i fork
tokio/src/process/mod.rs
616:    /// `fork`. This primarily means that any modifications made to memory on
620:    /// other threads perhaps still running when the `fork` was run).

tokio/tests/process_kill_on_drop.rs
18:       # Fork another child that won't get killed
```
As you can see, the only mention is in `process::pre_exec()` and a test for that mod.

While `fork()` after spawning threads is afaik not something broadly supported and could be somewhat expected as common knowledge, the `current_thread` runtime also doesn't work with `fork()` depending on features used.

As @Darksonn mentioned in Discord, `fork()`ing support in Tokio isn't well investigated, so including a call for help in the docs might be an option too.

That said, use cases for both fork and tokio seem either rare, or with an easy workaround:

- rare: severe isolation, like [esrlabs/northstar](https://github.com/esrlabs/northstar)
- easy workaround: daemonization for long running services: this is worked around by forking before running the tokio runtime

**Comment by Darksonn** [comment]
Fork is only supported in the following two cases:

1. The call to `fork` happens before the parent process uses Tokio in any way.
2. The child process does not touch Tokio in any way.

**Comment by correabuscar** [comment]
**tl;dr:** tokio could maybe use `pthread_atfork` to detect any illegally used `fork` attempts. If even feasible.

Perhaps, worst case, tokio could use [`pthread_atfork`](https://man7.org/linux/man-pages/man3/pthread_atfork.3.html) to install hooks([example](https://github.com/correabuscar/sandbox/tree/67dd7162d0c0aac4baaa34ee3a950c369c33e781/rust/05_sandbox/fork/pthread_atfork_try2) run `./t` or `./g`) to detect when `fork`-ing happens and thus `panic!` if it's used in the wrong cases. 

All targets that support `fork` support `pthread_atfork` which is these `#[cfg(any(unix, target_os = "fuchsia", target_os = "vxworks"))]`. From my superficial understanding:)

However, I'm not sure how feasible this is, because I've never used tokio yet, maybe some tokio parts can be used(called by user) without init-ing tokio first, thus these hooks would be useless then because they wouldn't have been installed yet.

Or, even if it can work, it's not worth the overhead of checking a `bool` on each possible tokio action to see if to panic, to satisfy this:
> The child process does not touch Tokio in any way.

Unless, any use of any tokio parts require some kind of tokio init to have happened first, then the check can be inside that init and thus be real simple.

Those hooks however, once set, will be run on every `fork` execution. But if they just set a `bool`, seems lightweight.

Just a thought, only tokio devs would know how feasible can this be.

**Comment by zbentley** [comment]
> Fork is only supported in the following two cases:
> 
>     1. The call to `fork` happens before the parent process uses Tokio in any way.
> 
>     2. The child process does not touch Tokio in any way.

That's good info, thanks.

Just to be excruciatingly clear: is  itunsupported to use a brand new Tokio runtime in a child process whose parent was running Tokio at the time of `fork`, provided that no code in the child process directly interacts with any Futures or Tokio-related objects created in the parent?

Specifically, I'm concerned about environments that make heavy use of forking that may be using Tokio, or even Rust, without knowing it.

Consider a Python project using any one of the large and growing number of Python libraries which implement native code in Rust, e.g. `py-polars` or `connector-x`. Python concurrency is frequently implemented via forking (`multiprocessing` in the stdlib does this, as well as many popular application server harnesses like Gunicorn). 

If such a Python project calls into a library whose native code uses Tokio, then Python later forks, and then the child Python fork calls into a different unit of native code that *also* uses Tokio, is that safe? If so, how can Tokio specifically be used on both sides of a fork such that this safety is achieved? If not, are all of these projects at risk of errors?

**Comment by Darksonn** [comment]
Yes. The problem is global variables that are per-runtime. Primarily, these have to do with handling signals and child processes. Forking breaks these globals.

> If such a Python project calls into a library whose native code uses Tokio, then Python later forks, and then the child Python fork calls into a different unit of native code that also uses Tokio, is that safe? If so, how can Tokio specifically be used on both sides of a fork such that this safety is achieved? If not, are all of these projects at risk of errors?

Depends on whether both pieces of native code use "the same Tokio". When shared libraries are involved, you can sometimes end up with multiple copies of Tokio and hence multiple copies of the global variables. Different copies of Tokio are independent.

**Comment by zbentley** [comment]
Interesting. Can those global PID registries/self-pipes/SIGCHLD handlers be reset in any way? Are they reset "by coincidence" when a new Runtime is constructed? Are they cleared by runtime shutdown?

Avoiding false sharing by library address space loading does likely mitigate the situation for things like Python, as you say, but it doesn't do much for Rust projects whose crates internally and independently use Tokio, nor does it help with Rust-linking projects that have rust at the top layer (an "open faced sandwich", I guess), e.g. "rust/tokio orchestration server embeds a Python interpreter in rust, Python interpreter loads unrelated Rust/tokio Python library, then Python forks and calls into that library".

**Comment by Darksonn** [comment]
No, there is currently no way to reset them.

**Comment by zbentley** [comment]
Understood. Would you review a doc PR summarizing what you've indicated in this discussion, or do you think this issue thread is sufficient?

**Comment by Darksonn** [comment]
I'm happy to see a PR document these details.

**Comment by Veeupup** [comment]
hi @Darksonn , thanks for your earlier replies and they are helpful!

What will happen in this case: I have process A which uses `tokio` as its runtime and then fork A to generate a child Process B. I will call `std::mem::forget(runtime)` in child process B forget the old runtime (which is forked from A) and create a new `tokio` runtime.

Will something bad happen in process A runtime or process B runtime?

**Comment by Darksonn** [comment]
Process B will not be able to use Tokio.

**Comment by igrigorik** [comment]
@Darksonn, could you elaborate on the complexity and work required to support the above use case? Specifically, what is the global state that makes the current implementation unsafe/incompatible with fork, and what would need to be true for this to be resolved? 

I'm not looking for a detailed breakdown. I'd like to get a high-level understanding of the hidden complexity and gauge potential appetite for addressing this.. if someone would be willing to do the work.

**Comment by Darksonn** [comment]
The process / signal modules define some global things that break on fork. For example, I think we sometimes have a global file descriptor for sending some notifications, and it doesn't survive the fork. There's also the signal handler, which may or may not behave well across a fork.

**Comment by igrigorik** [comment]
Thanks. If someone attempted a patch to tackle the above, is it something you'd support and entertain? Or are there some architecture/philosophy principles at play that would render it "it's just the way it is and forever will be".

**Comment by Darksonn** [comment]
I'm open to improving the situation. Shutting down all runtimes, calling `fork`, and then starting new runtimes afterwards seems possible to support.

**Comment by Xuanwo** [comment]
> I'm open to improving the situation. Shutting down all runtimes, calling `fork`, and then starting new runtimes afterwards seems possible to support.

Hi, @Darksonn, could you explain how to implement this? I'm happy to do the necessary work.

**Comment by Darksonn** [comment]
Well, there probably needs to be some mechanism to recreate the file descriptor mentioned above. But I don't really remember the details.
